### PR TITLE
VoxelGridOcclusionEstimation should always round down to go from coordinates to voxel indices.

### LINKED ISSUE #5941: [filters] VoxelGridOcclusionEstimation not working for some inputs
I have a sample point cloud in which occlusion estimation says nothing is occluded, regardless of voxel size, which is untrue. I am confused by this because it works on other point clouds.

**Context**

After the recent fixes to `VoxelGridOcclusionEstimation`, I have used this filter and it seemed to be working as expected, provided I set the voxel size big enough to actually occlude something. But now I ran across a sample where its obvious that some space is occluded, but everything is marked free.

A screenshot is here, the upper red dot marks the sensor origin in the cloud, the lower one is just 1 voxel I am casting to

image

**Expected behavior**

Many voxels occluded

**Current Behavior**

Nothing occluded

**To Reproduce**

Compile 910ccc61737612daad06e11fc79e51540d6480fc (current master), then run

```
./build/bin/pcl_voxel_grid_occlusion_estimation wat.pcd -leaf 0.2,0.2,0.2
``` 

or any other voxel size. The result is always like

```
> Ray-Traversal [done, 0.005125 ms : 0 occluded voxels]
```

**Your Environment (please complete the following information):**

 - OS: macos 12.6.5
 - Compiler: apple clang-1400.0.29.202
 - PCL Version master HEAD

**Possible Solution**

no idea

[wat.pcd.zip](https://github.com/PointCloudLibrary/pcl/files/14063525/wat.pcd.zip)

### PR DESCRIPTION
**Pull request by themightyoarfish**
Fixes #5941

### CODE REVIEW

**Comment by mvieth** [comment]
Please add the following test (e.g. in test/filters/test_filters.cpp):
```c++
TEST (VoxelGridOcclusionEstimation, Filters)
{
  auto input_cloud = pcl::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
  input_cloud->emplace_back(0.0, 0.0, 0.0);
  input_cloud->emplace_back(9.9, 9.9, 9.9); // we want a nice bounding box from (0, 0, 0) to (10, 10, 10)
  input_cloud->sensor_origin_ << -0.1, 0.5, 0.5; // just outside the bounding box. Most rays will enter at voxel (0, 0, 0)
  pcl::VoxelGridOcclusionEstimation<pcl::PointXYZ> vg;
  vg.setInputCloud (input_cloud);
  vg.setLeafSize (1.0, 1.0, 1.0);
  vg.initializeVoxelGrid ();
  std::vector<Eigen::Vector3i, Eigen::aligned_allocator<Eigen::Vector3i> > occluded_voxels;
  vg.occlusionEstimationAll (occluded_voxels);
  for(std::size_t y=0; y<10; ++y) {
    for (std::size_t z=0; z<10; ++z) {
      if(y==9 && z==9) continue; // voxel (9, 9, 9) is occupied by point (9.9, 9.9, 9.9), so it will not be counted as occluded
      Eigen::Vector3i cell(9, y, z);
      EXPECT_NE(std::find(occluded_voxels.begin(), occluded_voxels.end(), cell), occluded_voxels.end()); // not equal means it was found
    }
  }
}
```
And I found another problem: if the sensor origin is outside the bounding box, then `t_min` is computed so that `start` is right at the border of the bounding box. But because of numerical errors it can happen that `start` is slightly outside of the box, and then the computed grid coordinates `ijk` are outside of the box as well. Using `(t_min>0.0f?(t_min+std::numeric_limits<float>::epsilon()):t_min)` when computing `start` fixes this.

**Comment by themightyoarfish** [comment]
Updated the tests, passes locally. I added the other change, though I wonder if we can provoke it for test purposes.

File: filters/include/pcl/filters/impl/voxel_grid_occlusion_estimation.hpp [hunk 1]
```diff
@@ -351,8 +351,7 @@ pcl::VoxelGridOcclusionEstimation<PointT>::rayTraversal (std::vector<Eigen::Vect
   Eigen::Vector4f start = origin + t_min * direction;
```
Comment (mvieth) [inline filters/include/pcl/filters/impl/voxel_grid_occlusion_estimation.hpp:351]: Please adapt `t_min` here, like above

**Review by mvieth — COMMENTED** [review]
The new test should cover the added epsilon. At least on my computer I see a difference. Before (without adding epsilon):
Screenshot from 2024-01-30 10-40-51
Now (with the epsilon):
Screenshot from 2024-01-30 10-42-07
The sensor origin is behind the red voxel at the lower left. The 10x10 voxels facing the viewer should all be occluded (this is what is checked in the test).

File: filters/include/pcl/filters/impl/voxel_grid_occlusion_estimation.hpp (hunk 1 above)
Comment (themightyoarfish) [reply filters/include/pcl/filters/impl/voxel_grid_occlusion_estimation.hpp]: 99109f7c57500ae0b3d0e32f544de59b5993b092

**Review by mvieth — APPROVED** [review]
> When I run the tests with and without the change, they all pass regardless.

Might depend on compiler or compiler flags?

Thank you for the fix!

**Pull request merged**